### ISSUE TITLE
TVDBv2 Indexer API Issue

### ISSUE THREAD

**Issue by dotsam** [pymedusa/Medusa#11218]
The TVDBv2 API is currently returning bad data for Saturday Night Live (ID 76177) in a way which causes the SHOWQUEUE-UPDATE process to get stuck in a loop

https://github.com/pymedusa/Medusa/blob/6b37ad0ae6dfc3c2b102e2d9a377ced1fc849476/medusa/indexers/tvdbv2/api.py#L262-L284

When paging and reaching `https://api.thetvdb.com/series/76177/episodes/query?page=10`, the returned data I'm seeing is

```json
{
  "links": {
    "first": 1,
    "last": 13,
    "next": 11,
    "prev": 9
  },
  "data": null
}
```

With data being `null`, the `if paged_episodes.data:` conditional is never met, and thus the `page` and `last` variables are never updated, and the process just continues to loop.

Relevant chunk of a debug log:
```
2023-05-15 13:03:04 DEBUG    SHOWQUEUE-UPDATE :: [671a443] 76177: Beginning update of Saturday Night Live
2023-05-15 13:03:04 DEBUG    SHOWQUEUE-UPDATE :: [671a443] 76177: Retrieving show info from TVDBv2
2023-05-15 13:03:04 DEBUG    SHOWQUEUE-UPDATE :: [671a443] 76177: Using language from show settings: en
2023-05-15 13:03:04 DEBUG    SHOWQUEUE-UPDATE :: [671a443] 76177: Loading show info from TVDBv2
2023-05-15 13:03:04 DEBUG    SHOWQUEUE-UPDATE :: [671a443] Getting all series data for 76177
2023-05-15 13:03:04 DEBUG    SHOWQUEUE-UPDATE :: [671a443] Getting all show data for 76177
2023-05-15 13:03:05 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177 [Status: 200]
2023-05-15 13:03:05 DEBUG    SHOWQUEUE-UPDATE :: [671a443] Getting all episodes of 76177
2023-05-15 13:03:05 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=1 [Status: 200]
2023-05-15 13:03:05 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=2 [Status: 200]
2023-05-15 13:03:05 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=3 [Status: 200]
2023-05-15 13:03:05 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=4 [Status: 200]
2023-05-15 13:03:05 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=5 [Status: 200]
2023-05-15 13:03:05 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=6 [Status: 200]
2023-05-15 13:03:05 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=7 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=8 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=9 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
2023-05-15 13:03:06 DEBUG    SHOWQUEUE-UPDATE :: [671a443] GET URL: https://api.thetvdb.com/series/76177/episodes/query?page=10 [Status: 200]
... continues forever until restarted ...
```

**Comment by dotsam** [comment]
Looks like this was changed in #7846, which was a fix for trying to use null data.

**Comment by dotsam** [comment]
I haven't seen this causing CPU issues for me, but I also have CPU throttling disabled, which might help with that. This could also then be the cause of #11217 

In the short term, I've made the following change locally to fix this, but there's probably a better patch to be made to prevent other edge-case issues:
```diff
diff --git a/medusa/indexers/tvdbv2/api.py b/medusa/indexers/tvdbv2/api.py
index 44866ec88..4631c634a 100644
--- a/medusa/indexers/tvdbv2/api.py
+++ b/medusa/indexers/tvdbv2/api.py
@@ -270,8 +270,8 @@ class TVDBv2(BaseIndexer):
                         )
                         if paged_episodes.data:
                             results += paged_episodes.data
-                            last = paged_episodes.links.last
-                            page += 1
+                        last = paged_episodes.links.last
+                        page += 1
             else:
                 while page <= last:
                     paged_episodes = self.config['session'].series_api.series_id_episodes_query_get(
@@ -279,8 +279,8 @@ class TVDBv2(BaseIndexer):
                     )
                     if paged_episodes.data:
                         results += paged_episodes.data
-                        last = paged_episodes.links.last
-                        page += 1
+                    last = paged_episodes.links.last
+                    page += 1

             if results and full_info:
                 results = self._get_episodes_info(tvdb_id, results, season=aired_season)
```

**Comment by bassduh** [comment]
> In the short term, I've made the following change locally to fix this, but there's probably a better patch to be made to prevent other edge-case issues:
> 
At least the fix is stopping it from getting into an infinite loop, nice one!

**Comment by GldRush98** [comment]
Yeah, I tried that fix for a show I've ran in to this issue on; Chicago Med.
The fix does stop the loop, but it's also causing Medusa to not get any season information past S05E17. All episodes after that up through S08E20 are blank. 

It just so happens that S05E17 is the 100th episode in the series, and the last one pulling data. So it looks like TVDB's api returns info on 100 episodes at a time, and should be returning data on the next 100 episodes, but is just returning null.  Good grief.

I changed the Indexes from TVDB to TVMaze for this show and it pulled all episode info again just fine, so yeah, TVDB's busted api is the source of the problem.

**Comment by papampi** [comment]
I'm having same issue with showid 377183 and I dont know which show it is :
```
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] GET URL: https://api.thetvdb.com/series/388173/episodes/query?page=1 [Status: 200]
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] User-Agent: Medusa/1.0.13 (Linux; 5.10.60-qnap; 9b5a7de4-f324-11ed-8ff9-0e4aa9a86ce7)
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] GET URL: https://api.thetvdb.com/series/388173/episodes/query?page=1 [Status: 200]
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] User-Agent: Medusa/1.0.13 (Linux; 5.10.60-qnap; 9b5a7de4-f324-11ed-8ff9-0e4aa9a86ce7)
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] GET URL: https://api.thetvdb.com/series/388173/episodes/query?page=1 [Status: 200]
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] User-Agent: Medusa/1.0.13 (Linux; 5.10.60-qnap; 9b5a7de4-f324-11ed-8ff9-0e4aa9a86ce7)
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] GET URL: https://api.thetvdb.com/series/388173/episodes/query?page=1 [Status: 200]
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] User-Agent: Medusa/1.0.13 (Linux; 5.10.60-qnap; 9b5a7de4-f324-11ed-8ff9-0e4aa9a86ce7)
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] GET URL: https://api.thetvdb.com/series/388173/episodes/query?page=1 [Status: 200]
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] User-Agent: Medusa/1.0.13 (Linux; 5.10.60-qnap; 9b5a7de4-f324-11ed-8ff9-0e4aa9a86ce7)
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] GET URL: https://api.thetvdb.com/series/388173/episodes/query?page=1 [Status: 200]
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] User-Agent: Medusa/1.0.13 (Linux; 5.10.60-qnap; 9b5a7de4-f324-11ed-8ff9-0e4aa9a86ce7)
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] GET URL: https://api.thetvdb.com/series/388173/episodes/query?page=1 [Status: 200]
2023-05-16 11:02:12 DEBUG    SHOWUPDATER :: [6b37ad0] User-Agent: Medusa/1.0.13 (Linux; 5.10.60-qnap; 9b5a7de4-f324-11ed-8ff9-0e4aa9a86ce7)

```

Thousands and thousands of this line in the log and keeps going until I restart medusa

**Comment by GldRush98** [comment]
I created a support ticket on TheTVDB's ticket system for their API returning bad data.
I'm not going to hold my breath on them actually doing anything with the ticket though, solely based on past experiences with TVDB staff members being incredibly unpleasant to deal with.

**Comment by GldRush98** [comment]
Heard back from TVDB on the ticket. It sounds like they know it's broken and they don't care at all since it's an old API.  This is highly unlikely to be fixed on TVDB's side it sounds.

>Those endpoints are within our legacy API, which is no longer officially supported.  The documentation for our v4 API is available here: https://github.com/thetvdb/v4-api

If you run in to this bug, you should figure out what show is causing it via debug logging and change its indexer to anything other than TVDB. It might be time I change all of my show's indexers TBH.

**Comment by papampi** [comment]
I already changed all to TMDB a month ago and just found out 4 shows was missed during the migration to TMDB.

It's so much time consuming if you have large library since you cant select all and migrate them, you should select one by one and select from drop down which indexer you want to migrate to.

**Comment by dotsam** [comment]
There's actually been a bulk Indexer updater since [v0.5.23](https://github.com/pymedusa/Medusa/releases/tag/v0.5.23), it's under _Manage > Change Indexer_, and yes, it seems like now is probably the time to start thinking about switching away from TVDB

**Comment by papampi** [comment]
I'm on 1.0.13 but couldnt find how to use mass indexer change?
I clicked the check mark next to each show then select from drop down the indexer I want to migrate to for over 1000 shows, the single check mark on top do nothing.
Was there any other simpler way?


![photo1684299842](https://github.com/pymedusa/Medusa/assets/1108483/0a1f0b3b-e1cf-4e7b-8043-1312927ad7a1)

**Comment by spattersTatt** [comment]
I am also having this issue, with the loop causing CPU usuage to cause medusa to stop responding.

I tried the mass indexer change, but that was also painful as it cause medusa to start re-grabbing all the archove shows, and have the seasons were not listed.. so now going back to TVDB again and then monitor logs for the looping show :-(

**Comment by johnvick** [comment]
I have found the offending series so far to have been BBC documentaries and Panorama - both have hundreds of episodes. Shorter series seem OK with TVDB. I have deleted both - no other indexer available. Maybe no need to change all shows to new indexer.

**Comment by GldRush98** [comment]
I wouldn't leave any show on TVDB. Their API is broken, per their own admission, and since it's a legacy API, they aren't fixing it. You don't know what shows will randomly stop working tomorrow or next week or next month. Move off TVDB completely.

**Comment by johnvick** [comment]
I take your point, you're no doubt right. Any easy way to get banners when using TMDB?

**Comment by Minxster** [comment]
> I haven't seen this causing CPU issues for me, but I also have CPU throttling disabled, which might help with that. This could also then be the cause of #11217 
> 
> In the short term, I've made the following change locally to fix this, but there's probably a better patch to be made to prevent other edge-case issues:
> ```diff
> diff --git a/medusa/indexers/tvdbv2/api.py b/medusa/indexers/tvdbv2/api.py
> index 44866ec88..4631c634a 100644
> --- a/medusa/indexers/tvdbv2/api.py
> +++ b/medusa/indexers/tvdbv2/api.py
> @@ -270,8 +270,8 @@ class TVDBv2(BaseIndexer):
>                          )
>                          if paged_episodes.data:
>                              results += paged_episodes.data
> -                            last = paged_episodes.links.last
> -                            page += 1
> +                        last = paged_episodes.links.last
> +                        page += 1
>              else:
>                  while page <= last:
>                      paged_episodes = self.config['session'].series_api.series_id_episodes_query_get(
> @@ -279,8 +279,8 @@ class TVDBv2(BaseIndexer):
>                      )
>                      if paged_episodes.data:
>                          results += paged_episodes.data
> -                        last = paged_episodes.links.last
> -                        page += 1
> +                    last = paged_episodes.links.last
> +                    page += 1
> 
>              if results and full_info:
>                  results = self._get_episodes_info(tvdb_id, results, season=aired_season)
> ```

Any chance of pushing this, as I'm currently setting up Medusa in docker and trying to migrate from SickChill. Bulk library imports are impossible as everything has to default to TVDB until it's in Medusa? So the import fails drastically 🤷‍♂️

**Comment by medariox** [comment]
I just released a new version (1.0.14) that contains @dotsam's code: https://github.com/pymedusa/Medusa/releases/tag/v1.0.14
Thanks again for the fix!

You are still advised to migrate away from TVDB if possible.

**Comment by jxer** [comment]
> I just released a new version (1.0.14) that contains @dotsam's code: https://github.com/pymedusa/Medusa/releases/tag/v1.0.14 Thanks again for the fix!
> 
> You are still advised to migrate away from TVDB if possible.

Can you provide a path for us to do this? the bulk select check box doesn't work, and apparently moving to a different indexer causes shows to flip back to wanted status, can you help here @medariox my one box has 214 shows, with 5k+ episodes in archive mode

**Comment by jxer** [comment]
> > > I just released a new version (1.0.14) that contains @dotsam's code: https://github.com/pymedusa/Medusa/releases/tag/v1.0.14 Thanks again for the fix!
> > > You are still advised to migrate away from TVDB if possible.
> > 
> > 
> > Can you provide a path for us to do this? the bulk select check box doesn't work, and apparently moving to a different indexer causes shows to flip back to wanted status, can you help here @medariox my one box has 214 shows, with 5k+ episodes in archive mode
> 
> It doesn't cause to flip to Wanted. It depends what is selected as your default show status. Change the default show status to Archived and try with one show. When you are confident it works as you want it to, you can migrate with more shows.

ok i will do that, is it possible to fix the select all option in there? it doesn't work

**Comment by jxer** [comment]
maybe I am missing something @medariox , why are we not simply moving to the v4 api coding?

**Comment by medariox** [comment]
> > > > I just released a new version (1.0.14) that contains @dotsam's code: https://github.com/pymedusa/Medusa/releases/tag/v1.0.14 Thanks again for the fix!
> > > > You are still advised to migrate away from TVDB if possible.
> > > 
> > > 
> > > Can you provide a path for us to do this? the bulk select check box doesn't work, and apparently moving to a different indexer causes shows to flip back to wanted status, can you help here @medariox my one box has 214 shows, with 5k+ episodes in archive mode
> > 
> > 
> > It doesn't cause to flip to Wanted. It depends what is selected as your default show status. Change the default show status to Archived and try with one show. When you are confident it works as you want it to, you can migrate with more shows.
> 
> ok i will do that, is it possible to fix the select all option in there? it doesn't work

No, it's not. It doesn't make sense to have a select all option, because you should do all shows at once in the first place. If you really want to, you need to invest 5 minutes of your time to click the shows you want to migrate.

**Comment by Minxster** [comment]
Thanks for updates @medariox , I'll check in on docker versions shortly to see if they're ready to pull down. If in doubt I may try a temporary fix inside docker and stop any repulls (or host reboots 🤣).

Just on the "things switch to 'wanted'", I saw this with some of my early small-batch tests. There is a problem with it doing that and I'm worried about my full bulk library import, but... I also saw that once shows are full refreshed a file rescan happens (afterwards), so my plan is to just disable any torrent/bin connections until the process has finished doing it's thing... I'm in no rush to download, just want to get Medusa as shipshape as possible first 😉

**Comment by Minxster** [comment]
I just wanted to confirm that update at least lets me run imports ok, so I should be able to work through the rest. Thanks @medariox

**Comment by mmhere** [comment]
I know this is closed but I wanted to confirm that the changes to indexers/tvdbv2/api.py (both in 1.0.14 and 1.0.15, with the latter being the latest) fixed this for me.

I had been seeing saturated CPU, plus a memory leak that eventually killed the entire Linux machine on which it was running (swapping+thrashing and eventual death).

Thanks!

P.S.:  I'm gradually moving all shows from thetvdb's old/broken API to tvmaze, which will take some time with >600 shows that I want to do manually and observe to ensure nothing goes awry with the more obscure shows.  Probly oughta cull some dead shows at the same time ;-)

**Comment by bobpaul** [comment]
> I wouldn't leave any show on TVDB. Their API is broken, per their own admission, and since it's a legacy API, they aren't fixing it. You don't know what shows will randomly stop working tomorrow or next week or next month. Move off TVDB completely.

So I understand moving away from TVDB is a work around, but why isn't the push here to migrate pymedusa to their newer API that they still support?

**Comment by jxer** [comment]
> > I wouldn't leave any show on TVDB. Their API is broken, per their own admission, and since it's a legacy API, they aren't fixing it. You don't know what shows will randomly stop working tomorrow or next week or next month. Move off TVDB completely.
> 
> So I understand moving away from TVDB is a work around, but why isn't the push here to migrate pymedusa to their newer API that they still support?

per another thread, they are charging money for it

**Comment by mmhere** [comment]
(cross post from 11217;  apologies for cross-posting.)

A couple of years ago there was a huge discussion over here https://github.com/pymedusa/Medusa/discussions/8738 about thetvdb *planning* to monetize their latest API, and what the response here might be.

I'll append my comments there so as not to pollute this issue.

TL;DR over there:

https://github.com/pymedusa/Medusa/discussions/8738#discussioncomment-5962127
